dtm0/fop.c

```diff
@@ -200,6 +206,76 @@ static size_t dtm0_fom_locality(const struct m0_fom *fom)
 	return locality++;
 }
 
+static void m0_dtm0_post__persistent(struct m0_dtm0_service *dtms,
+				     const struct m0_fid *tgt,
+				     const struct m0_dtm0_tx_desc *txd)
+{
+	struct m0_fop          *fop;
+	struct m0_rpc_session  *session;
+	struct m0_rpc_item     *item;
+	struct dtm0_req_fop    *req;
+	int                     rc;
+
+	M0_ENTRY("reqh=%p", dtms->dos_generic.rs_reqh);
+
+	session = m0_dtm0_service_process_session_get(&dtms->dos_generic, tgt);
+
+	fop               = m0_fop_alloc_at(session, &dtm0_req_fop_fopt);
+	item              = &fop->f_item;
+	item->ri_ops      = &dtm0_req_fop_rpc_item_ops;
+	item->ri_session  = session;
+	item->ri_prio     = M0_RPC_ITEM_PRIO_MID;
+	item->ri_deadline = M0_TIME_IMMEDIATELY;
+
+	req               = m0_fop_data(fop);
+	req->dtr_msg      = DTM_PERSISTENT;
+	rc = m0_dtm0_tx_desc_copy(txd, &req->dtr_txr);
+	M0_ASSERT(rc == 0);
+
+	rc = m0_rpc_post(item);
+	M0_ASSERT(rc == 0);
+	m0_fop_put_lock(fop);
+
+	M0_LEAVE("Sent PERSISTENT notice " FID_F " -> " FID_F,
+		 FID_P(&dtms->dos_generic.rs_service_fid), FID_P(tgt));
+}
+
+
+M0_INTERNAL void m0_dtm0_on_persistent(struct m0_reqh               *reqh,
+				       const struct m0_dtm0_tx_desc *txd)
+{
+	int                     rc;
+	struct m0_dtm0_service *dtms;
+	struct m0_dtm0_tx_desc  msg = {};
+	int                     i;
+
+	dtms = m0_dtm0_service_find(reqh);
+
+	rc = m0_dtm0_tx_desc_copy(txd, &msg);
+	M0_ASSERT(rc == 0);
+
+	/* TODO: This change will be done by DTM0 log because
+	 * it should log the entry "pa == self" with PERSISTENT state
+	 * set.
+	 */
+	for (i = 0; i < msg.dtd_pg.dtpg_nr; ++i) {
+		if (m0_fid_eq(&msg.dtd_pg.dtpg_pa[i].pa_fid,
+			      &dtms->dos_generic.rs_service_fid)) {
+			msg.dtd_pg.dtpg_pa[i].pa_state =
+				max_check(msg.dtd_pg.dtpg_pa[i].pa_state,
+					  (uint32_t) M0_DTPS_PERSISTENT);
+		}
+	}
+
+	/* Notify the originator */
+	m0_dtm0_post__persistent(dtms, &msg.dtd_id.dti_fid, &msg);
+
+	/* TODO: Send notices to the rest of the participants. */
+	m0_dtm0_tx_desc_fini(&msg);
+}
+
+
+
 #define M0_FID(c_, k_)  { .f_container = c_, .f_key = k_ }
 static void dtm0_pong_back(enum m0_be_dtm0_log_credit_op msgtype,
```

Comment:
m0_dtm0_post__persistent and dtm0_pong_back are similar, what about to reduce copy paste?